<h1>Grip@ The Sparks Foundation<br><br>
Data Science and Business Analytics Intern</h1>

<h2>Sanskriti Bansal</h2>

<h2>Prediction Using Unsupervised Learning</h2>

### Importing modules

In [2]:
import numpy as np
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

### Importing dataset

In [3]:
df=pd.read_csv("Live.csv")

In [4]:
df.head(10)

status_id status_type status_published  \
0  246675545449582_1649696485147474       video   4/22/2018 6:00   
1  246675545449582_1649426988507757       photo  4/21/2018 22:45   
2  246675545449582_1648730588577397       video   4/21/2018 6:17   
3  246675545449582_1648576705259452       photo   4/21/2018 2:29   
4  246675545449582_1645700502213739       photo   4/18/2018 3:22   
5  246675545449582_1645650162218773       photo   4/18/2018 2:14   
6  246675545449582_1645564175560705       video   4/18/2018 0:24   
7  246675545449582_1644824665634656       video   4/17/2018 7:42   
8  246675545449582_1644655795651543       photo   4/17/2018 3:33   
9  246675545449582_1638788379571618       photo   4/11/2018 4:53   

   num_reactions  num_comments  num_shares  num_likes  num_loves  num_wows  \
0            529           512         262        432         92         3   
1            150             0           0        150          0         0   
2            227           236          57        204         21         1   
3            111             0           0        111          0         0   
4            213             0           0        204          9         0   
5            217             6           0        211          5         1   
6            503           614          72        418         70        10   
7            295           453          53        260         32         1   
8            203             1           0        198          5         0   
9            170             9           1        167          3         0   

   num_hahas  num_sads  num_angrys  Column1  Column2  Column3  Column4  
0          1         1           0      NaN      NaN      NaN      NaN  
1          0         0           0      NaN      NaN      NaN      NaN  
2          1         0           0      NaN      NaN      NaN      NaN  
3          0         0           0      NaN      NaN      NaN      NaN  
4          0         0           0      NaN      NaN      NaN      NaN  
5          0         0           0      NaN      NaN      NaN      NaN  
6          2         0           3      NaN      NaN      NaN      NaN  
7          1         0           1      NaN      NaN      NaN      NaN  
8          0         0           0      NaN      NaN      NaN      NaN  
9          0         0           0      NaN      NaN      NaN      NaN

### Exploratory data analysis

In [3]:
df.shape

(7050, 16)

In [4]:
df.isnull().sum()

status_id              0
status_type            0
status_published       0
num_reactions          0
num_comments           0
num_shares             0
num_likes              0
num_loves              0
num_wows               0
num_hahas              0
num_sads               0
num_angrys             0
Column1             7050
Column2             7050
Column3             7050
Column4             7050
dtype: int64

In [5]:
# Since all the tuples in column1, column2, column3, column4 are null, 
# these attributes can be dropped

df.drop(['Column1', 'Column2', 'Column3', 'Column4'], axis=1, inplace=True)

In [6]:
df.shape

(7050, 12)

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7050 entries, 0 to 7049
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   status_id         7050 non-null   object
 1   status_type       7050 non-null   object
 2   status_published  7050 non-null   object
 3   num_reactions     7050 non-null   int64 
 4   num_comments      7050 non-null   int64 
 5   num_shares        7050 non-null   int64 
 6   num_likes         7050 non-null   int64 
 7   num_loves         7050 non-null   int64 
 8   num_wows          7050 non-null   int64 
 9   num_hahas         7050 non-null   int64 
 10  num_sads          7050 non-null   int64 
 11  num_angrys        7050 non-null   int64 
dtypes: int64(9), object(3)
memory usage: 661.1+ KB


In [8]:
df.head()

status_id status_type status_published  \
0  246675545449582_1649696485147474       video   4/22/2018 6:00   
1  246675545449582_1649426988507757       photo  4/21/2018 22:45   
2  246675545449582_1648730588577397       video   4/21/2018 6:17   
3  246675545449582_1648576705259452       photo   4/21/2018 2:29   
4  246675545449582_1645700502213739       photo   4/18/2018 3:22   

   num_reactions  num_comments  num_shares  num_likes  num_loves  num_wows  \
0            529           512         262        432         92         3   
1            150             0           0        150          0         0   
2            227           236          57        204         21         1   
3            111             0           0        111          0         0   
4            213             0           0        204          9         0   

   num_hahas  num_sads  num_angrys  
0          1         1           0  
1          0         0           0  
2          1         0           0  
3          0         0           0  
4          0         0           0

In [9]:
df.describe()

num_reactions  num_comments   num_shares    num_likes    num_loves  \
count    7050.000000   7050.000000  7050.000000  7050.000000  7050.000000   
mean      230.117163    224.356028    40.022553   215.043121    12.728652   
std       462.625309    889.636820   131.599965   449.472357    39.972930   
min         0.000000      0.000000     0.000000     0.000000     0.000000   
25%        17.000000      0.000000     0.000000    17.000000     0.000000   
50%        59.500000      4.000000     0.000000    58.000000     0.000000   
75%       219.000000     23.000000     4.000000   184.750000     3.000000   
max      4710.000000  20990.000000  3424.000000  4710.000000   657.000000   

          num_wows    num_hahas     num_sads   num_angrys  
count  7050.000000  7050.000000  7050.000000  7050.000000  
mean      1.289362     0.696454     0.243688     0.113191  
std       8.719650     3.957183     1.597156     0.726812  
min       0.000000     0.000000     0.000000     0.000000  
25%       0.000000     0.000000     0.000000     0.000000  
50%       0.000000     0.000000     0.000000     0.000000  
75%       0.000000     0.000000     0.000000     0.000000  
max     278.000000   157.000000    51.000000    31.000000

In [10]:
# view the labels in the variable status_id

df['status_id'].unique()

array(['246675545449582_1649696485147474',
       '246675545449582_1649426988507757',
       '246675545449582_1648730588577397', ...,
       '1050855161656896_1060126464063099',
       '1050855161656896_1058663487542730',
       '1050855161656896_1050858841656528'], dtype=object)

In [11]:
len(df['status_id'].unique())

6997

There are 6997 unique labels in the status_id variable. The total number of instances in the dataset is 7050. So, it is approximately a unique identifier for each of the instances. Thus this is not a variable that can be used. Hence, it can be dropped.

In [12]:
# view the labels in the variable status_published

df['status_published'].unique()

array(['4/22/2018 6:00', '4/21/2018 22:45', '4/21/2018 6:17', ...,
       '9/21/2016 23:03', '9/20/2016 0:43', '9/10/2016 10:30'],
      dtype=object)

In [13]:
# view how many different types of variables are there

len(df['status_published'].unique())

6913

There are 6913 unique labels in the status_published variable. The total number of instances in the dataset is 7050. So, it is also approximately a unique identifier for each of the instances. Thus this is not a variable that can be used. Hence, it can be dropped.

In [14]:
# view the labels in the variable status_type

df['status_type'].unique()

array(['video', 'photo', 'link', 'status'], dtype=object)

In [15]:
# view how many different types of variables are there

len(df['status_type'].unique())

4

In [16]:
df.drop(['status_id', 'status_published'], axis=1, inplace=True)

In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7050 entries, 0 to 7049
Data columns (total 10 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   status_type    7050 non-null   object
 1   num_reactions  7050 non-null   int64 
 2   num_comments   7050 non-null   int64 
 3   num_shares     7050 non-null   int64 
 4   num_likes      7050 non-null   int64 
 5   num_loves      7050 non-null   int64 
 6   num_wows       7050 non-null   int64 
 7   num_hahas      7050 non-null   int64 
 8   num_sads       7050 non-null   int64 
 9   num_angrys     7050 non-null   int64 
dtypes: int64(9), object(1)
memory usage: 550.9+ KB


In [18]:
df.head()

status_type  num_reactions  num_comments  num_shares  num_likes  num_loves  \
0       video            529           512         262        432         92   
1       photo            150             0           0        150          0   
2       video            227           236          57        204         21   
3       photo            111             0           0        111          0   
4       photo            213             0           0        204          9   

   num_wows  num_hahas  num_sads  num_angrys  
0         3          1         1           0  
1         0          0         0           0  
2         1          1         0           0  
3         0          0         0           0  
4         0          0         0           0

### Converting categorical variable into integers

In [19]:
X = df
Y = df['status_type']

In [20]:
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()
X['status_type'] = le.fit_transform(X['status_type'])
Y = le.transform(Y)

In [21]:
X.head()

status_type  num_reactions  num_comments  num_shares  num_likes  num_loves  \
0            3            529           512         262        432         92   
1            1            150             0           0        150          0   
2            3            227           236          57        204         21   
3            1            111             0           0        111          0   
4            1            213             0           0        204          9   

   num_wows  num_hahas  num_sads  num_angrys  
0         3          1         1           0  
1         0          0         0           0  
2         1          1         0           0  
3         0          0         0           0  
4         0          0         0           0

### Feature scaling

In [22]:
cols = X.columns

In [23]:
from sklearn.preprocessing import MinMaxScaler

ms = MinMaxScaler()
X = ms.fit_transform(X)

In [24]:
X = pd.DataFrame(X, columns=[cols])
X.head()

status_type num_reactions num_comments num_shares num_likes num_loves  \
0    1.000000      0.112314     0.024393   0.076519  0.091720  0.140030   
1    0.333333      0.031847     0.000000   0.000000  0.031847  0.000000   
2    1.000000      0.048195     0.011243   0.016647  0.043312  0.031963   
3    0.333333      0.023567     0.000000   0.000000  0.023567  0.000000   
4    0.333333      0.045223     0.000000   0.000000  0.043312  0.013699   

   num_wows num_hahas  num_sads num_angrys  
0  0.010791  0.006369  0.019608        0.0  
1  0.000000  0.000000  0.000000        0.0  
2  0.003597  0.006369  0.000000        0.0  
3  0.000000  0.000000  0.000000        0.0  
4  0.000000  0.000000  0.000000        0.0

### Model implementation

In [26]:
from sklearn.cluster import KMeans

kmeans = KMeans(n_clusters=4, random_state=0)

kmeans.fit(X)
Ypred = kmeans.fit_predict(X)

# To check how many of the samples are correctly labeled
labels = kmeans.labels_

correct_labels = sum(Y == labels)
print("Result: %d out of %d samples were correctly labeled." % (correct_labels, Y.size))
print('Accuracy score: {0:0.2f}'. format(correct_labels/float(Y.size)))

Result: 4340 out of 7050 samples were correctly labeled.
Accuracy score: 0.62


### Performance measurement

In [27]:
# Accuracy score

from sklearn import metrics
print("Accuracy:",metrics.accuracy_score(Y, Ypred))

Accuracy: 0.6156028368794326


In [28]:
# Classification report

from sklearn.metrics import classification_report
print(classification_report(Y,Ypred))

              precision    recall  f1-score   support

           0       0.00      0.00      0.00        63
           1       0.99      0.95      0.97      4288
           2       0.25      0.22      0.24       365
           3       1.00      0.08      0.15      2334

    accuracy                           0.62      7050
   macro avg       0.56      0.31      0.34      7050
weighted avg       0.94      0.62      0.65      7050

